core/message_handler.go

```diff
@@ -32,13 +46,17 @@ func (app *App) GetMessagesToShare(max int) ([][]byte, error) {
 	if count == 0 {
 		return nil, nil
 	}
-	// If count is less than max, we use an offset of 0 to simply return all
-	// orders we have stored.
-	offset := 0
-	if count > max {
-		// If count is greater than max, choose an offset such that we always return
-		// at least max orders.
-		offset = rand.Intn(count - max)
+	// The offset will be the value closest to `nextOffset` such that the
+	// maximum number of orders that can be shared are selected.
```

Comment:
Hmm, I think there is a small inefficiency in this implementation. Suppose `count = 25` orders and `max = 10`:

1. On the first iteration, `nextOffset = 0` and `offset = 0`. We select orders 0 through 9. Then we set `nextOffset = 10`.
2. On the second iteration, `nextOffset = 10` and `offset = 10`. We select orders 10 through 19. Then we set `nextOffset = 20`.
3. On the third iteration, `nextOffset = 20`. Here the condition `count - offset < max` is true, so we would set `offset = max(count - max, 0) = 15`. We would select orders 15 through 24.

Step 3 is inefficient since we already shared orders 15 through 19 in the previous iteration.

Instead of this, maybe we could either:

1. Leave `offset` unchanged and select orders 20 through 24 in step 3. It isn't hugely important to always select `max` orders now that we don't have to worry about the probability of some orders never being selected.
2. Perform a second query and select orders 20 through 24 *and* orders 0 through 4. Then set `nextOffset = 5`. This is the most optimal version but adds a little bit of complexity with the second query.

Reply:
I liked the second strategy, so that's what I ended up going with. It does increase the complexity, but I don't think that it's too bad.